### PR TITLE
new lint: unused_unit

### PR DESCRIPTION
**Pull request by llogiq**
This is the rejected unused_parens extension plus a lint for units in function return types and return expressions.

It appears to fail to obey the `allow(..)` attributes though (see `copies.rs`) – does anyone know why that happens?

### CODE REVIEW

**Comment by flip1995** [comment]
Maybe the ignored `allow(..)` is related to #3198. Something weird is going on with defining lint levels. I didn't find the time to investigate this further yet.

**Comment by llogiq** [comment]
Maybe the lint level selection was broken when tools lints were added?

**Comment by flip1995** [comment]
I thought the same, so I searched for merge commits by @bors that broke the lint level definition and I got it down to 3 commits: https://github.com/rust-lang-nursery/rust-clippy/issues/3198#issuecomment-423607115. None of these commits are related to `tool_lints` though..

**Comment by llogiq** [comment]
I have extended the lint to cover explicit unit return types in `fn` types and `Fn*` trait bounds. Apart from the lint level selection issue, this should be ready to merge.

**Comment by llogiq** [comment]
I had missed adding the lint to the lint-array, which resulted in the failing allow annotation. This should hopefully now work. r? once CI is happy.

File: tests/ui/unused_unit.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+// run-rustfix
+// compile-pass
+
+// The output for humans should just highlight the whole span without showing
+// the suggested replacement, but we also want to test that suggested
+// replacement only removes one set of parentheses, rather than naïvely
+// stripping away any starting or ending parenthesis characters—hence this
+// test of the JSON error format.
+#![feature(tool_lints)]
```
Comment (oli-obk) [inline tests/ui/unused_unit.rs]: this is not needed anymore, tool lints are stable

**Comment by llogiq** [comment]
@oli-obk r?

File: clippy_lints/src/returns.rs [hunk 2]
```diff
@@ -68,6 +68,25 @@ declare_clippy_lint! {
      the end of a block"
 }
 
+/// **What it does:** Checks for unit (`()`) expressions that can be removed
```
Comment (mati865) [inline clippy_lints/src/returns.rs]: Maybe too pedantic but there is no dot at the end of the line.

**Comment by aldanor** [comment]
@llogiq This seems to act weird in cases like this:

```rust
thread_local!(static A: () = ());
```

yields

```
warning: unneeded unit expression
   |
20 | thread_local!(static A: () = ());
   |                              ^^ help: remove the final `()`
```

... and `allow(clippy::unused_unit)` doesn't seem to work in this case, for some reason.

**Comment by llogiq** [comment]
True, that looks weird. Then again, why do you need a zero-sized thread-local?

**Pull request merged**